keps/sig-node/1870-provisioning-resources-with-numa-topology/README.md

```diff
@@ -0,0 +1,278 @@
+
+To calculate available resources need to know all resources
+which could be used by kubernetes. It could be calculated by
+subtracting resources of kube cgroup and system cgroup from
+system allocatable resources.
+
+
+### Design based on podresources interface of the kubelet
+
+Podresources interface of the kubelet is described in
+
+[pkg/kubelet/apis/podresources/v1alpha1/api.proto](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/podresources/v1alpha1/api.proto)
+
+it is available for every process on the worker node by
+unix domain socket situated by the following path:
+
+```go
+filepath.Join(kl.getRootDir(), config.DefaultKubeletPodResourcesDirName)
+```
+
+it could be used to collect used resources on the worker node and to evaluate
+its NUMA assignment (by device id).
+
+Podresources could also be used to obtain initial information on resources of the worker node.
+
+The PodResource API as it stands today:
+* only provides information from Device Manager but not from CPU Manager.
+* doesn't contain topology information as part of ContainerDevice.
+* doesn't have the capability to let clients enumerate the resources.
+
+This [KEP](https://github.com/kubernetes/enhancements/pull/1884) proposes extension of podresource api to address the above mentioned gaps.
+
+With the changes proposed in the above KEP, this interface might look like as following:
+
+```proto
+syntax = "proto3";
+
+package v1alpha1;
+
+
+service PodResources {
+    rpc List(ListPodResourcesRequest) returns (ListPodResourcesResponse) {}
+    rpc GetAvailableResources(AvailableResourcesRequest) returns (AvailableResourcesResponse) {}
+}
+
+message ListPodResourcesRequest {}
+
+message ListPodResourcesResponse {
+    repeated PodResources pod_resources = 1;
+}
+
+message AvailableResourcesRequest {}
+
+message AvailableResourcesResponse {
+    repeated ContainerDevices devices = 1;
+    repeated int64 cpu_ids = 2;
+}
+
+message ContainerDevices {
+    string resource_name = 1;
+    repeated string device_ids = 2;
+    Topology topology = 3;
+}
+```
+
+### API
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit).
+
+
+```go
+// NodeResourceTopology is a specification for a Foo resource
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	TopologyPolicies []string `json:"topologyPolicies"`
+	Zones            ZoneMap  `json:"zones"`
+}
+
+// Zone is the spec for a NodeResourceTopology resource
+type Zone struct {
+	Type       string           `json:"type"`
+	Parent     string           `json:"parent,omitempty"`
+	Costs      map[string]int   `json:"costs,omitempty"`
+	Attributes map[string]int   `json:"attributes,omitempty"`
+	Resources  ResourceInfoMap  `json:"resources,omitempty"`
```

Comment:
Is there any chance inconsistency in scheduling because for filtering we make use  of NodeResourceTopology  whereas for scoring resourceAllocationScorer can be used which is based on nodeinfo ? 
I hope we can't ensure data consistency  in both NodeResourceTopology  and nodeinfo as these are updated from different processes.

Reply:
This structure was intended to keep node hardware topology, to be used in https://github.com/kubernetes/enhancements/pull/1858. Kube-scheduler plugin keeps the instances of this structure. So it's not yet in NodeInfo and wasn't planed for the first stage (alpha), and there is no relation between NodeInfo  and NodeResourceTopology.
Gap of inconsistency in kube-scheduler could exist for small amount of time, this inconsistency is bigger than inconsistency with whole resources in NodeInfo, e.g. when it used in NodeResourcesFit. Once again NodeResourceTopology is not going to be used in NodeResourceFit, it will be used in special plugin.

Reply:
Yes, I understand the filtering is going to be based on NodeResourceTopology in a special plugin, is scoring also to be based on  NodeResourceTopology ? AFAIK, it would still use ResourceAllocationScorer which is based on nodeinfo. Am I right here ?

Reply:
It depends on whether @AlexeyPerevalov intends to provide a Filter plugin only, or both Filter and Score plugins.

Reply:
We have been thinking about Score plugin too. Here is a document where we described the problem: https://docs.google.com/document/d/1dq_XE6h4ZN9Kxaxm4O0D_TKllFAOEjBkcBnVLLQT6Hc/edit
But decided to focus on Filter plugin at first stage. Score plugin will be added later, I think it worth mentioning it in KEP )